src/analysis/processing/qgsalgorithmroundrastervalues.cpp

```diff
@@ -0,0 +1,262 @@
+
+  QgsRasterIterator iter( mInterface.get() );
+  iter.startRasterRead( mBand, mLayerWidth, mLayerHeight, mExtent );
+  int iterLeft = 0;
+  int iterTop = 0;
+  int iterCols = 0;
+  int iterRows = 0;
+  std::unique_ptr< QgsRasterBlock > analysisRasterBlock;
+  while ( iter.readNextRasterPart( mBand, iterCols, iterRows, analysisRasterBlock, iterLeft, iterTop ) )
+  {
+    if ( feedback )
+      feedback->setProgress( 100 * ( ( iterTop / maxHeight * nbBlocksWidth ) + iterLeft / maxWidth ) / nbBlocks );
+    if ( mIsInteger && mDecimalPrecision > -1 )
+    {
+      //nothing to round, just write raster block
+      analysisRasterBlock->setNoDataValue( mInputNoDataValue );
+      destinationRasterProvider->writeBlock( analysisRasterBlock.get(), mBand, iterLeft, iterTop );
+    }
+    else
+    {
+      for ( int row = 0; row < iterRows; row++ )
+      {
+        if ( feedback && feedback->isCanceled() )
+          break;
+        for ( int column = 0; column < iterCols; column++ )
+        {
+          if ( analysisRasterBlock->isNoData( row, column ) )
+          {
+            analysisRasterBlock->setValue( row, column, mInputNoDataValue );
+          }
+          else
+          {
+            double val = analysisRasterBlock->value( row, column );
+            double roundedVal = mInputNoDataValue;
+            if ( mRoundingDirection == 0 && mDecimalPrecision < 0 )
+            {
+              roundedVal = roundUpBaseN( val, mMultipleOfBaseN );
+            }
+            else if ( mRoundingDirection == 0 && mDecimalPrecision > -1 )
+            {
+              roundedVal = roundUp( val, mDecimalPrecision );
+            }
+            else if ( mRoundingDirection == 1 && mDecimalPrecision < 0 )
+            {
+              roundedVal = roundNearestBaseN( val, mMultipleOfBaseN );
+            }
+            else if ( mRoundingDirection == 1 && mDecimalPrecision > -1 )
+            {
+              roundedVal = roundNearest( val, mDecimalPrecision );
+            }
+            else if ( mRoundingDirection == 2 && mDecimalPrecision < 0 )
+            {
+              roundedVal = roundDownBaseN( val, mMultipleOfBaseN );
+            }
+            else
+            {
+              roundedVal = roundDown( val, mDecimalPrecision );
+            }
+            //integer values get automatically cast to double when reading and back to int when writing
+            analysisRasterBlock->setValue( row, column, roundedVal );
+          }
+        }
+      }
+      destinationRasterProvider->writeBlock( analysisRasterBlock.get(), mBand, iterLeft, iterTop );
+    }
+  }
+  destinationRasterProvider->setEditable( false );
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), outputFile );
+  return outputs;
+}
+
+double QgsRoundRasterValuesAlgorithm::roundNearest( double &value, int &decimals )
+{
+  double m = ( value < 0.0 ) ? -1.0 : 1.0;
+  double scaleFactor = std::pow( 10.0, decimals );
```

Comment:
This scalefactor is getting calculated once per pixel -- it'd be preferable to calculate it once in processAlgorithm and pass it to this function.

Reply:
fixed

Reply:
Sorry, I think you misunderstood me -- it's "scale factor" which is a constant value and should only be calculated once